### ISSUE TITLE
[Request] Background Blinking

### ISSUE THREAD

**Issue by cgarwood** [bitfocus/companion#674]
**Describe the feature**
The ability to set the background and/or foreground colors to blink between two sets of colors. Bonus if we can configure the speed at which it changes.

**Is this platform dependent (windows, mac, ..)?**
no

**If documentation is required to implement, do you know where to find it?**
n/a

**Usecases**
If you're recording or streaming and want to draw more attention to that fact, you could have the background of the recording button blink between black and red.

Blinking the button for an input that's currently live on a switcher

Drawing attention to something that's in an error state.

etc.

**Comment by DachsbauTV** [comment]
Just somer more usecase examples:

- a muted microfon (never forget to unmute a commentator again after a break)
- a timer went to 0:00
- callback from an other position or device
- Battery level feedback
- on buttons with a lot of delayed steps it could be a great addition to see when it reached the last step

**Comment by plasticpool** [comment]
I agree with all the use cases here – just knowing that a microphone channel is open (or not) with a flashing/blinking button state would be super helpful, and I'd be happy to contribute to a bounty toward implementing this.

**Comment by mdgnys** [comment]
A great way to accomplish this would be a module that supplies instance feedback that is simply a constantly switching Boolean back and fourth.

**Comment by shaunar2006** [comment]
This would be awesome! I need this!

**Comment by justingiffard** [comment]
This would really be awesome, from a dev point of view it would be cool if there was a way to specify the frequency of the flashes too. I'd like to add feedback to the X32 module for the tap delay that it gives the user a visual idea of what the delay is like the tap delay on the console

**Comment by HaxNobody** [comment]
Here's how I implemented blinking buttons:
1. Set up a new custom variable, let's call it "Flash" for now. You should be able to use the same variable for multiple different buttons, and you can create more variables for different flash patterns.
2. Set up a trigger that is also called Flash or something similar, and configure the Condition to be Time Interval
3. Set the time interval to 1 (seconds)
4. Tick the "Relative Delays" box
5. Add an Action for "internal: Set custom variable value", configure the Delay to 500ms, select your custom variable, and set the Value to 0
6. Add another Action the exact same as step 5, but set the Value to 1
7. Configure your button to add a Feedback of "internal: Check variable value" and set it to your custom variable, with default '=' Operation and the Value as 1.
8.  Finally, configure the text color and background color that you want to switch to with each flash, different from the initial setting at the top of the button config.

Done - you can adjust all the trigger timings to make your flasher work the way you want it to.

**Comment by codered11343** [comment]
> Here's how I implemented blinking buttons:
> 
> 1. Set up a new custom variable, let's call it "Flash" for now. You should be able to use the same variable for multiple different buttons, and you can create more variables for different flash patterns.
> 2. Set up a trigger that is also called Flash or something similar, and configure the Condition to be Time Interval
> 3. Set the time interval to 1 (seconds)
> 4. Tick the "Relative Delays" box
> 5. Add an Action for "internal: Set custom variable value", configure the Delay to 500ms, select your custom variable, and set the Value to 0
> 6. Add another Action the exact same as step 5, but set the Value to 1
> 7. Configure your button to add a Feedback of "internal: Check variable value" and set it to your custom variable, with default '=' Operation and the Value as 1.
> 8. Finally, configure the text color and background color that you want to switch to with each flash, different from the initial setting at the top of the button config.
> 
> Done - you can adjust all the trigger timings to make your flasher work the way you want it to.

Thank you. This was exactly the guidance I was looking for. But to go a step further, would it be possible to make it so that button paused its flashing after the button was pressed, then resumed flashing after it hadn't been pressed after a certain amount of time? Like 5 minutes? If this is currently possible, how would you program that?

**Comment by dnmeid** [comment]
I think I just had an idea how we could implement this request in a flexible and useful way.
At the moment we have the checkbox to invert boolean feedback. We could replace this checkbox by a dropdown with some additional feedback behaviors:
* normal
* inverted
* temporary on on
* temporary on off
* temporary on change
* blinking slow
* blinking medium
* blinking fast

So users have control of what element should blink, eg. background color or an image or text.
Users will not have total control over blink frequency, instead all blinks are done in synchronized groups, leading to a less distractive surface. Maybe we make the three frequencies customizable in user preferences.

Current checkbox state can be upgraded to normal and inverted behavior.

**Comment by Julusian** [comment]
I'm not convinced that this should replace the invert checkbox, but having something in that left panel between that and the `change style properties` field would make sense.
![image](https://github.com/bitfocus/companion/assets/1327476/d626599f-e172-416f-9125-500f6990ac20)

My reason being that the `invert` box is part of the boolean feedbacks, which also gets used in the conditions for triggers, and eventually will be part of #1052 and #372, and having to handle a checkbox in some places and a dropdown in others will make everything more complicated.  
And I think if it is combined there would be more dropdown options too, such as `blinking slow inverted` for when you want it to be blinking when the feedback is false. everything would probably need an 'inverted' version, at which point having a separate checkbox might be simpler to understand.

**Comment by sackboydad** [comment]
> Here's how I implemented blinking buttons:
> 
> 1. Set up a new custom variable, let's call it "Flash" for now. You should be able to use the same variable for multiple different buttons, and you can create more variables for different flash patterns.
> 2. Set up a trigger that is also called Flash or something similar, and configure the Condition to be Time Interval
> 3. Set the time interval to 1 (seconds)
> 4. Tick the "Relative Delays" box
> 5. Add an Action for "internal: Set custom variable value", configure the Delay to 500ms, select your custom variable, and set the Value to 0
> 6. Add another Action the exact same as step 5, but set the Value to 1
> 7. Configure your button to add a Feedback of "internal: Check variable value" and set it to your custom variable, with default '=' Operation and the Value as 1.
> 8. Finally, configure the text color and background color that you want to switch to with each flash, different from the initial setting at the top of the button config.
> 
> Done - you can adjust all the trigger timings to make your flasher work the way you want it to.

Hi!  Hope you're still responding to this thread.   I implemented your procedure and it's great!  But, how can I make a button blinking conditional?  For example, I have a button that mutes a channel on my audio mixer.  Right now I use a feedback that is designed to just change the background color if muted.  Using your procedure how can I turn that into a blinking button?   Thanks!

**Comment by Donaldini** [comment]
@sackboydad 
You can change the Trigger to add a conditional. Your trigger would then no longer be universal for everything, but it works.

I just created a button that will blink whenever i'm injecting too much solar power into the grid and another for when my electric cars battery falls below a certain percentage encouraging me to connect it to the charger.

@HaxNobody thanks for this tutorial!

**Comment by sackboydad** [comment]
> @sackboydad You can change the Trigger to add a conditional. Your trigger would then no longer be universal for everything, but it works.
> 
> I just created a button that will blink whenever i'm injecting too much solar power into the grid and another for when my electric cars battery falls below a certain percentage encouraging me to connect it to the charger.
> 
> @HaxNobody thanks for this tutorial!

Thank you.  I’m not the best on Companion is there any chance you have the time to show me how to put what you suggested in companion?  Thank you very much

**Comment by Donaldini** [comment]
Sure, here's a screenshot of my Trigger:

<img width="1737" alt="image" src="https://github.com/bitfocus/companion/assets/2185680/f4a23584-d240-4e6c-bb58-faf68d70ef16">

I used a _Check boolean expression_ because I needed to cast a variable coming from my Home Assistant environment to a Number (it was a string). The rest of the expression is to check whether the surplus value of my home electrical grid is below the value -4. This means that i'm using a negative of 4kW of electricity in my house OR simply put i'm injecting more than 4kW to the grid instead of using it. _Time to do some laundry or charge the car_ ;-)

**Comment by sackboydad** [comment]
> Sure, here's a screenshot of my Trigger:
> 
> <img alt="image" width="1737" src="https://private-user-images.githubusercontent.com/2185680/326202184-f4a23584-d240-4e6c-bb58-faf68d70ef16.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQyNTU3MTYsIm5iZiI6MTcxNDI1NTQxNiwicGF0aCI6Ii8yMTg1NjgwLzMyNjIwMjE4NC1mNGEyMzU4NC1kMjQwLTRlNmMtYmI1OC1mYWY2OGQ3MGVmMTYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDQyNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA0MjdUMjIwMzM2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDkxODQ0NmEwZGQxNjljZGU0NTk4YjU2YTk5ZjE4NWExMTQzNGUyNzhiNDA1ZjQzNTU5NjY0NTZlYTA2MzRlMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.KF4jKds-tiBLZaonyUaftp175rAi376V6AqfEeeRb2Q">
> I used a _Check boolean expression_ because I needed to cast a variable coming from my Home Assistant environment to a Number (it was a string). The rest of the expression is to check whether the surplus value of my home electrical grid is below the value -4. This means that i'm using a negative of 4kW of electricity in my house OR simply put i'm injecting more than 4kW to the grid instead of using it. _Time to do some laundry or charge the car_ ;-)

One more question... I PROMISE I'm not going to become a leech bugging you!  Once I get pointed in the right direction I'm pretty good......  We use a WING audio mixer, whose Companion module isn't the best.   It does have a published exhaustive set of OSC commands.  I have successfully used the customer command in a button on Companion to issue a OSC command.    Looking at your settings, I see that I would use the trigger condition to decide if something blinks.  But when I try to set a condition on our WING mixer, it says there aren't any.   I see you used an expression of some kind to see about your charger level.  Would there be a way to use an OSC command or inquiry in the trigger condition?   Here is an example of an OSC command that would either set or show the fader level of Bus1's Fader.  How would you put that (or can you) into a trigger condition?  I SWEAR no more questions!   

OSC:    /bus/1/fdr  (the number range is between -144 and +10

Thank you!

**Comment by Donaldini** [comment]
No problem at all, happy to help.

I don't have a Behringer Wing, but when I look at the documentation inside companion I see that there are some dynamic variabels and feedback variables available that might help you. See here:

![image](https://github.com/bitfocus/companion/assets/2185680/1e35f2ee-322b-465a-9e07-5f5c8340221c)

<img width="781" alt="image" src="https://github.com/bitfocus/companion/assets/2185680/a22ba47d-1686-4983-9d8f-d0c2a14a385f">

________________________________________________________________________________________________________

I think you need to use one of these variables: 

- $(INSTANCENAME:{Strip}_d)
- $(INSTANCENAME:{Strip}_{Bus}_d)

I'm not entirely sure I understand your question correctly, but here's a go:
For example if you want to use that level to determine whether or not to let a button blink you could use my expression and change it to: _+$(INSTANCENAME:{Strip}_d) > -10_ (for letting it blink when it gets too loud)

**Comment by sackboydad** [comment]
Thank you so much.  Even if this doesn't work I have learned a whole lot!  I will try it out.  True to my word, I won't bug you any more about this, short of letting you know if it worked.

**Comment by whopkinscom** [comment]
I use generic-blink module to achieve this.
![image](https://github.com/bitfocus/companion/assets/17707253/cd343db4-4d0f-4c34-a3ed-7228e143ffc6)
For example When I change a camera input on ATEM I reset and start a timer. The feedback on my camera buttons will utilise generic-blink if the timer has passed say 2 minutes to remind me I haven't switched cameras in 2 mins so maybe I want to think about that?
![image](https://github.com/bitfocus/companion/assets/17707253/b9d99534-f57f-4f46-9a86-c75567d3db1c)

**Comment by NoSloppy** [comment]
> I use generic-blink module to achieve this.

I don't understand how to set this up. Could you explain? Thanks. I don't find generic blink anywhere

**Comment by Coupe70** [comment]
@josephdadams 
Are you the developer of generic-blink?
I didn't find a seperate github for it and don't know where else to post...
Would it be possible to get an action "blink button / don't blink button" in addition to the feedback functionality?
For buttons without external feedback it's not 100% convenient to create a custom variable for each button
that might be supposed to blink and set it with an action just to get the blinking running via feedback...

**Comment by cpuguru07** [comment]
Y'all are overcomplicating this. No need for triggers or custom variables.

To create a button that flashes every second:
Feedback: internal: Variable: Check Boolean expression
Expression: $(internal:time_s)%2==0

Combine with internal: Logic: Operation to create conditional flashing.

**Comment by sackboydad** [comment]
Or you could just use the Button Blinker module.   It’s nearly perfect.  I’ve been using it for almost 2 years.

**Comment by thedist** [comment]
> Or you could just use the Button Blinker module. It’s nearly perfect. I’ve been using it for almost 2 years.

A native Companion solution, such as one using Expressions, will always be a better workflow than having to rely on a module (which if such a module was made today it may not have even been included in Companion).

Just because the module does work for basic use cases doesn't mean that it should be the recommended option.

**Comment by sackboydad** [comment]
> > Or you could just use the Button Blinker module. It’s nearly perfect. I’ve been using it for almost 2 years.
> 
> A native Companion solution, such as one using Expressions, will always be a better workflow than having to rely on a module (which if such a module was made today it may not have even been included in Companion).
> 
> Just because the module does work for basic use cases doesn't mean that it should be the recommended option.

I agree. I just really like how easy it is to use the module.  Different iterations of module let me easily change my blink rate, etc.  I like that not only does it let me blink but I can have one feedback set but gives me a lot of flexibility.   

For instance, for many of the buttons on my panels, I want to show an “on” state, an “off” state and an “error” state.   So on the button, I set font size, text, BKGD and color on the button definition.   Then with a feedback, evaluated for an “off” state or an error state where it’s, say, red BKGD, yellow text, size 15.  But if there’s an error state, in addition to the last button change (off or error), I use the blinker module.  If error then i change the text size to 8, BKGD to black, text to white and a 750ms blink rate.  So in an error the button not only blinks, but the text size pulsates in size and changes color to really attract attention.

**Comment by Julusian** [comment]
An idea I should look into as part of the graphics overhaul (or maybe before then);
Maybe there could/should be an expression function to provide this blinking.

Perhaps something like:
```
blink(700) // possibly allow the interval to be optional, with a default value from somewhere?
```

internally, companion would probably do some variable magic to make this work while keeping them all in sync

This doesnt solve my concerns about the cost of constantly re-rendering, but maybe I am being too optimistic about wanting a solution to do that..

**Comment by arikorn** [comment]
Just cross-referencing Julian's PR #3831, which implements what @Julusian described in the previous comment.

**Comment by Julusian** [comment]
As hinted at above, in 4.3 there is now a `blink(500)` expression function.

I think I would recommend defining yourself some blink expression variables, to give yourself some easy references so that you only need to define the rate in one place. But this function will work anywhere where expressions can be used, so you can define it however you like

**Comment by sackboydad** [comment]
> As hinted at above, in 4.3 there is now a `blink(500)` expression function.
> 
> I think I would recommend defining yourself some blink expression variables, to give yourself some easy references so that you only need to define the rate in one place. But this function will work anywhere where expressions can be used, so you can define it however you like

Hey Julian .....  this is great!  thanks.   Question.... is there more than one argument in the expression?  When I type "blink" in the expression editor the template it gives me has 2 arguments but if I put two in it just stops blinking.  Could the second number be (please, please) the amount of time to remain off before blinking again?  I.e. just having 500 gives an even blink.  But, using GenericBlinkerModule, sometimes I don't want a huge blink but I want someone to see a slight "nudge". So I have a different on and off time in that module.  For instance, Canon Camera Auto-tracking button.... it blinks for 2000ms, then blinks off for 250ms.  That way it's MOSTLY lit, but just does a short quick un-blink for 250ms then goes back to a 2000ms blink.  Hope that makes sense.

<img width="476" height="90" alt="Image" src="https://github.com/user-attachments/assets/1e12cf3b-b1a5-4bd5-8078-bed1a8703b70" />

**Comment by arikorn** [comment]
@sackboydad , good point -- the description could be more ... descriptive.

The arguments are different than in generic-blink: the first arg is the total period, the second is fraction on a number from 0 - 1. So blink(1000, 500) is always on.  Thinking on it, I suppose we could change the behavior depending on greater than or less than 1, but I hesitate to recommend it, as it could result in very confusing expressions.

I'll fix the doc-string in a PR.  But here's the documentation (you have to select beta on the website, but it should be in the bundled docs you can access from the help menu (upper-right question-mark menu): https://companion.free/user-guide/beta/expressions/functions#:~:text=blink(number%2C%20%3Fnumber)

also there's a "hidden" more-details caret that shows up at the right of the "hints" menu when you hover over the item. Clicking on it shows (the line spacing is a bit much, but the content does scroll):

<img width="763" height="338" alt="Image" src="https://github.com/user-attachments/assets/8e1b6b51-1311-4e6b-a44e-fdb276ae0efe" />

my suggestion:

<img width="440" height="63" alt="Image" src="https://github.com/user-attachments/assets/6c0fd195-724e-42b9-93a3-4075bab4c3dd" />

**Comment by sackboydad** [comment]
OK, I get it.  So the second number is a fraction of the first number, where that fraction is the "percentage" of time the blink color is ON compared to the blink OFF.   So blink(3000,.9) would have each cycle take 3 seconds, with 90% blink OFF and 10% blink ON?    I like it.  I'm sure it's got to take less system resources than the module, right?

EDIT: (removed question about using in an expression.  I figured it out)

**Comment by arikorn** [comment]
> I'm sure it's got to take less system resources than the module, right?

That's almost certainly true - modules run in separate processes, so consume extra system resources.

> So blink(3000,.9) would have each cycle take 3 seconds, with 90% blink OFF and 10% blink ON?

You got it! And as for your deleted question, the `&&` operator is your friend. Glad you worked it out.

**Comment by sackboydad** [comment]
Thanks Ari.   Now the big wait is for 4.3 release (or is it 4.4?).  Much as I love this I can’t risk other parts of the beta in my environment
